Src/PerformanceCollector/Shared/Implementation/QuickPulse/Process/PerfLib/PerformanceMonitor.cs

```diff
@@ -0,0 +1,109 @@
+        /// we wait may not be sufficient if the Win32 code keeps running into this deadlock again 
+        /// and again. A condition very rare but possible in theory. We would get back to the user 
+        /// in this case with InvalidOperationException after the wait time expires.
+        /// </summary>
+        public byte[] GetData(string categoryIndex)
+        {
+            int waitRetries = 3; // 17;   //2^16*10ms == approximately 10mins
+            int waitSleep = 0;
+            int error = 0;
+
+            // no need to revert here since we'll fall off the end of the method
+            new RegistryPermission(PermissionState.Unrestricted).Assert();
+            while (waitRetries > 0)
+            {
+                try
+                {
+                    return (byte[])this.perfDataKey.GetValue(categoryIndex);
+                }
+                catch (IOException e)
+                {
+                    error = Marshal.GetHRForException(e);
+                    switch (error)
+                    {
+                        case NativeMethods.RPC_S_CALL_FAILED:
+                        case NativeMethods.ERROR_INVALID_HANDLE:
+                        case NativeMethods.RPC_S_SERVER_UNAVAILABLE:
+                            this.Init();
+                            goto case NativeMethods.WAIT_TIMEOUT;
+
+                        case NativeMethods.WAIT_TIMEOUT:
+                        case NativeMethods.ERROR_NOT_READY:
+                        case NativeMethods.ERROR_LOCK_FAILED:
+                        case NativeMethods.ERROR_BUSY:
+                            --waitRetries;
+                            if (waitSleep == 0)
+                            {
+                                waitSleep = 10;
+                            }
+                            else
+                            {
+                                Thread.Sleep(waitSleep);
+                                waitSleep *= 2;
+                            }
+
+                            break;
+
+                        default:
+                            throw SharedUtils.CreateSafeWin32Exception(error);
+                    }
+                }
+                catch (InvalidCastException e)
+                {
+                    throw new InvalidOperationException("Counter data is corrupt " + this.perfDataKey, e);
+                }
+            }
+
+            throw SharedUtils.CreateSafeWin32Exception(error);
+        }
+
+        [ResourceExposure(ResourceScope.None)]
+        [ResourceConsumption(ResourceScope.Machine, ResourceScope.Machine)]
+        private void Init()
+        {
+            try
+            {
+                this.perfDataKey = Registry.PerformanceData;
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                throw new UnauthorizedAccessException("Access denied opening the performance key", e);
```

Comment:
Why wrap the UAE with another UAE?  Wrapping and rethrowing should provide additional value which it isn't clear if that is the case here.  What is the original UAE message?